qrand/platforms/qsharp/circuit.py

```diff
@@ -28,76 +28,129 @@
+            raise ValueError(
+                f"Qubit index out of range {qubit_index} >= {self.num_qubits}."
+            )
 
     @property
     def num_qubits(self) -> int:
-        raise NotImplementedError(self.ERROR_MSG)
+        return self._num_qubits
 
     ############################## SPECIAL GATES ##############################
     def measure(self, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"""
+        if M(q[{target_qubit}]) == One{{
+        set value w/= {target_qubit} <- "1";
+        }}
+        """
 
     ########################### SINGLE QUBIT GATES ###########################
     def h(self, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"H(q[{target_qubit}]);"
 
     def rx(self, radians: float, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"Rx({radians},q[{target_qubit}]);"
 
     def ry(self, radians: float, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"Ry({radians},q[{target_qubit}]);"
 
     def rz(self, radians: float, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"Rz({radians},q[{target_qubit}]);"
 
     def s(self, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"S(q[{target_qubit}]);"
 
     def t(self, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"T(q[{target_qubit}]);"
 
     def u1(self, theta: float, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"R1({theta},q[{target_qubit}]);"
 
     def u2(self, phi: float, lam: float, target_qubit: int) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"""
+        Rz({lam},q[{target_qubit}]);
+        Ry(0.5*PI(),q[{target_qubit}]);
+        Rz({phi},q[{target_qubit}]);
+        """
 
     def u3(
         self, theta: float, phi: float, lam: float, target_qubit: int
     ) -> None:
-        raise NotImplementedError(self.ERROR_MSG)
+        self._validate_qubit_index(target_qubit)
+        self.gates += f"""
+        Rz({lam},q[{target_qubit}]);
+        Rx(-0.5*PI(),q[{target_qubit}]);
+        Rz({theta},q[{target_qubit}]);
+        Rx(0.5*PI(),q[{target_qubit}]);
+        Rz({phi},q[{target_qubit}]);
```

Comment:
These `RX` are in reversed order. See [here](https://qiskit.org/documentation/stubs/qiskit.circuit.library.U3Gate.html) you start applying the gates from right to left.

It should be:
```
Rz({lam},q[{target_qubit}]);
Rx(0.5*PI(),q[{target_qubit}]);
Rz({theta},q[{target_qubit}]);
Rx(-0.5*PI(),q[{target_qubit}]);
Rz({phi},q[{target_qubit}]);
```

Reply:
Fixed.